MobilityData/mobility-feed-api
# feat: add support for storing and validating GBFS files in GCP storage

**Summary:**

Closes #566 by adding support for storing GBFS files from an auto-discovery URL in a Cloud Storage bucket. The updates include the creation of a new `gbfs.json` file with paths pointing to the stored files in the bucket, validating the feeds, and storing the necessary information in the database. The changes involve the creation of the `gbfs-validator-batch` and `gbfs-validator-pubsub` functions to support this new functionality, ensuring that GBFS feeds are properly validated and stored.

**Expected behavior:**

This pull request introduces the following behavior:
- The `gbfs-validator-batch` function triggers an execution that checks all GBFS feeds in the database and publishes a message to the Pub/Sub topic for each feed.
- The `gbfs-validator-pubsub` function processes these messages, downloads feed snapshots to GCP, validates them, saves the validation report summary to GCP, and updates the database with the snapshot information and validation report.

**Testing tips:**

1. Trigger the `batch` function:
   - Execute the following command to trigger the batch function:
     ```bash
     curl -m 550 -X POST https://northamerica-northeast1-mobility-feeds-dev.cloudfunctions.net/gbfs-validator-batch \
     -H "Authorization: bearer $(gcloud auth print-identity-token)" \
     -H "Content-Type: application/json"
     ```
   - Alternatively, trigger the Cloud Scheduler by selecting [`gbfs-validator-batch-scheduler-dev`](https://console.cloud.google.com/cloudscheduler?project=mobility-feeds-dev) and clicking `Force Run`.

2. The `gbfs-validator-pubsub` function should be triggered automatically. You can verify the logs to ensure everything is functioning correctly: [Pub/Sub Function Logs](https://console.cloud.google.com/functions/details/northamerica-northeast1/gbfs-validator-pubsub?env=gen2&project=mobility-feeds-dev&tab=logs).

3. Check the stored feeds in the specified Cloud Storage bucket. Ensure the `gbfs.json` file and the validation report are correctly generated: [GBFS Storage Bucket](https://console.cloud.google.com/storage/browser/mobilitydata-gbfs-snapshots-dev;tab=objects?forceOnBucketsSortingFiltering=true&project=mobility-feeds-dev&prefix=&forceOnObjectsSortingFiltering=false).

4. Verify that the database content in the DEV env reflects the expected changes and updates.

Testers are invited to follow the tips AND to try anything they deem relevant outside the bounds of the testing tips. 

Please make sure these boxes are checked before submitting your pull request - thanks!
